nemo/collections/tts/modules/submodules.py

```diff
@@ -401,3 +413,280 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        self.style_attention = StyleAttention(gru_hidden=gru_hidden, gst_size=gst_size, n_style_token=n_style_token, n_style_attn_head=n_style_attn_head)
+
+    @property
+    def input_types(self):
+        return {
+            "inp":NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inp_mask":NeuralType(('B', 'T_spec', 1), TokenDurationType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "gst": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inp, inp_mask):
+        style_embedding = self.reference_encoder(inp, inp_mask)
+        gst = self.style_attention(style_embedding)
+        return gst
+
+
+class ReferenceEncoderUtteranceLevel(NeuralModule):
+    def __init__(self, cnn_filters=[32, 32, 64, 64, 128, 128], dropout=0.2, gru_hidden=128):
+        super(ReferenceEncoderUtteranceLevel, self).__init__()
+        self.filter_size = [1] + cnn_filters
+        self.dropout = dropout
+        self.conv = torch.nn.Sequential(
+            OrderedDict(
+                [
+                    module
+                    for i in range(len(cnn_filters))
+                    for module in (
+                        (
+                            "conv2d_{}".format(i + 1),
+                            Conv2d(
+                                in_channels=int(self.filter_size[i]),
+                                out_channels=int(self.filter_size[i + 1]),
+                                kernel_size=(3, 3),
+                                stride=(2, 2),
+                                padding=(1, 1),
+                            ),
+                        ),
+                        ("relu_{}".format(i + 1), torch.nn.ReLU()),
+                        (
+                            "layer_norm_{}".format(i + 1),
+                            torch.nn.LayerNorm(self.filter_size[i + 1]),
+                        ),
+                        ("dropout_{}".format(i + 1), torch.nn.Dropout(self.dropout)),
+                    )
+                ]
+            )
+        )
+
+        self.gru = torch.nn.GRU(
+            input_size=cnn_filters[-1] * 2,
+            hidden_size=gru_hidden,
+            batch_first=True,
+        )
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inputs_masks": NeuralType(('B', 'T_spec', 1), TokenDurationType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "out": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs, inputs_masks):
+        inputs = inputs.transpose(1,2)
+
+        inputs = inputs * inputs_masks
+        out = inputs.unsqueeze(3)
+        out = self.conv(out)
+        out = out.view(out.shape[0], out.shape[1], -1).contiguous()
+        self.gru.flatten_parameters()
+        _, out = self.gru(out)
```

Comment:
It looks like masking in GRU requires converting its input to torch.nn.utils.rnn.pack_padded_sequence. Otherwise result will be very wrong for all but the longest sequence in the batch.

https://pytorch.org/docs/stable/generated/torch.nn.GRU.html?highlight=gru#torch.nn.GRU